### PR TITLE
:tada: effort to generate mnist data using GAN with tinygrad. [WIP]

### PR DESCRIPTION
**Pull request by kartik4949**
An effort to generate mnist digits with tinygrad.
Added generator
added LeakyRelu.
added dropout.

### CODE REVIEW

File: examples/mnist_gan.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+sys.path.append(os.getcwd())
+sys.path.append(os.path.join(os.getcwd(), 'test'))
+
+from tinygrad.tensor import Tensor, Function, register
+from tinygrad.utils import get_parameters
+import tinygrad.optim as optim
+
+GPU = os.getenv("GPU") is not None
+
+class LeakyReLU(Function):
```
Comment (geohot) [inline examples/mnist_gan.py:13]: If
```
x.relu() + (-leaky_slope*x).relu()
```
works I'd merge this as a second class op

Comment (kartik4949) [reply examples/mnist_gan.py:13]: let me check

Comment (kartik4949) [reply examples/mnist_gan.py:13]: and u mean first class op right.

Comment (geohot) [reply examples/mnist_gan.py:13]: (would need a test too)

Comment (kartik4949) [reply examples/mnist_gan.py:13]: yea ofc

Comment (kartik4949) [reply examples/mnist_gan.py:13]: worked

File: examples/mnist_gan.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+sys.path.append(os.getcwd())
+sys.path.append(os.path.join(os.getcwd(), 'test'))
+
+from tinygrad.tensor import Tensor, Function, register
+from tinygrad.utils import get_parameters
+import tinygrad.optim as optim
+
+GPU = os.getenv("GPU") is not None
+
+class Dropout(Function):
```
Comment (kartik4949) [inline examples/mnist_gan.py:13]: @geohot very needed ..

Comment (geohot) [reply examples/mnist_gan.py:13]: This def doesn't have to be a first class op (with a forward and backward). It's just a multiply by a mask.

Comment (kartik4949) [reply examples/mnist_gan.py:13]: But for backward you need same mask 
How u will generate those mask again to multiply with gradients

Comment (kartik4949) [reply examples/mnist_gan.py:13]: Oh so u mean making it second class
Like relu

Comment (kartik4949) [reply examples/mnist_gan.py:13]: These second class are funnn lol

**Comment by kartik4949** [comment]
worrking code waiting for training completion will post generated pictures.

**Comment by kartik4949** [comment]
@geohot 
losses are making sense now..
generated some images of '0' digit when overfitted with zero only train data
but with mode collapse.
TODO:
l2 reg
label smoothing
BCE

please check and help if you find some tweaks or bugs.
